lld/ELF/SyntheticSections.cpp

```diff
@@ -667,6 +667,20 @@ GotSection::GotSection(Ctx &ctx)
 void GotSection::addConstant(const Relocation &r) { relocations.push_back(r); }
 void GotSection::addEntry(const Symbol &sym) {
   assert(sym.auxIdx == ctx.symAux.size() - 1);
+  auto *d = dyn_cast<Defined>(&sym);
+  if (d && ctx.arg.icf != ICFLevel::None) {
+    // There may be symbols that have been ICFed in which case d->section
+    // points to their canonical section and d->value is offset in to that section.
+    // We add only a single GOT entry for all such symbols.
+    auto [it, inserted] = gotEntries.insert(
+      std::make_pair(std::make_pair(d->section, d->value),
+      numEntries));
+    if (!inserted && d->folded) {
```

Comment:
Is there a case when a `d->folded` is seen first and inserted into `gotEntries` then `d->folded == false` the canonical definition is encountered and we skip it?

I'm not sure if there's a good one-pass solution to this. Possibly if the value of the map were <Defined, uint32_t> then we could look at the symbol to see if it were folded.

Reply:
Yes, I'm not sure why we even need the `folded` check here. I think we can just let the first symbol we see set the `gotIdx`.